cloudsync/cs.py

```diff
@@ -247,38 +243,34 @@ def start(self, *, daemon=True, **kwargs):
         """
         Starts the cloudsync service.
         """
-        # Replaces :py:meth:`cloudsync.Runnable.start` so that events are processed asynchronously.
-        self.sthread = threading.Thread(target=self.smgr.run, kwargs={'sleep': 0.1, **kwargs}, daemon=daemon)
-        self.ethreads = (
-            threading.Thread(target=self.emgrs[0].run, kwargs={'sleep': self.sleep[0], **kwargs}, daemon=daemon),
-            threading.Thread(target=self.emgrs[1].run, kwargs={'sleep': self.sleep[1], **kwargs}, daemon=daemon)
-        )
-        self.sthread.start()
-        self.ethreads[0].start()
-        self.ethreads[1].start()
-        log.debug('Starting the notification manager')
+        log.debug("starting sync: %s", self.storage_label())
         self.nmgr.notify(Notification(SourceEnum.SYNC, NotificationType.STARTED, None))
-        self.nmgr.start(**kwargs)
-
+        self.smgr.start(daemon=daemon, sleep=0.1, **kwargs)
+        self.emgrs[0].start(daemon=daemon, sleep=self.sleep[0], **kwargs)
+        self.emgrs[1].start(daemon=daemon, sleep=self.sleep[1], **kwargs)
+        self.nmgr.start(daemon=daemon, **kwargs)
 
     def stop(self, forever=True, wait=True):
         """
         Stops the cloudsync service.
 
         Args:
             forever: If false is passed, then handles are left open for a future start.  Generally used for tests only.
+            wait: If false, manager threads are signaled to stop, but are NOT joined
         """
         log.info("stopping sync: %s", self.storage_label())
-        self.smgr.stop(forever=forever, wait=wait)
-        self.emgrs[0].stop(forever=forever, wait=wait)
-        self.emgrs[1].stop(forever=forever, wait=wait)
-        self.nmgr.stop(forever=forever, wait=wait)
-        if self.sthread:
-            self.sthread.join()
-            self.ethreads[0].join()
-            self.ethreads[1].join()
-            self.nmgr.wait()  # TODO: wait() above instead of join()?
-            self.sthread = None
+        for mgr in self._runnables:
+            mgr.stop(forever=forever, wait=wait)
+
+    @staticmethod
+    def stop_all(syncs: List["CloudSync"]):
```

Comment:
This belongs in the `Runnable` class

Reply:
are you suggesting Runnable should have a `stop_all(list[Runnable])` ?

if so, that would defeat the purpose, to an extent. This impl can stop N CloudSyncs efficiently (signal the N*4 runnables to stop, then wait for the N*4 runnables). Whereas calling `stop_all()` on only the Runnables that this CloudSync knows about would signal those 4 to stop, then wait for them, then the caller would need to do the same for all the other CloudSyncs in a loop, so you end up doing N iterations of (stop 4, wait 4)

Unless I am misunderstanding your comment?

Reply:
If you move `stop_all` to be a static method in `Runnable` you can take the list of runnables to stop as the argument to that method, and then do the same looping you have here, but do it there, on the list that was passed in. The cleanup of the `run` loop in `Runnable` should itself call `Runnable.stop_all` on any "sub-runnables" that are in that class, but the important thing is that calling `stop_all` on the "sub-runnables" happens in the *run thread* not in the main thread, so that way, if there are 10 CloudSyncs (which is itself a runnable with 4 sub-runnables) then all 50 runnables are spinning down simultaneously, and there are no calls to `stop` that are waiting on a call to `wait` related to some other call to `stop`. All 50 `stop`s happen, then all 50 `wait`s.

This will require adding a `run` method to cloudsync, but all that method would need to do is to hang waiting on some event forever, then `stop_all` on the sub-runnables, like:
```
SHUTTING_DOWN = Event()
def run()
	SHUTTING_DOWN.wait()
	Runnable.stop_all(self._runnables)

def stop()
	SHUTTING_DOWN.set()
```

Reply:
I think the cleanest solution might be to create a standard interface for sub-runnables, so that if a runnable spins up any runnables of its own, then it adds references to those runnables to an object property that maintains the set of sub-runnables, for the purpose of calling them all in the run thread, when the run thread drops out. Slightly more ambitious, perhaps, but if we have enough time to get this change in with a nice structure like that, it could be nicer

Reply:
Baking this structure into the `Runnable` class would be helpful, because it will be tempting to call `stop_all` in the `def stop()` of a runnable, which will be very inefficient. As we spoke about, what we're looking for is for `def stop()` to be very quick, set a flag and no real work, long-running operations or waiting

Reply:
So doing shutdown in a thread like this will work, but it takes away all the convenience of the convenience function, IMO, because the caller will need to wait on each cloudsync's shutdown to complete.

Reply:
As I mentioned, the design goal is to allow the 50 runnable threads from 10 cloudsyncs to all spin down simultaneously. If the method I described requires waiting on one cloudsync's shutdown to complete before beginning the shutdown of another cloudsync, then please explain what the flaw is in the design. Perhaps I'm mistaken, but I'm just not seeing it.

Reply:
performance-wise there is no flaw, my comment was related to the convenience aspect.

even without a `stop_all()` func the client can simply do this to shutdown multiple cloudsyncs efficiently:

```
for cs in cloudsyncs:
  cs.stop(wait=False)
for cs in cloudsyncs:
  cs.wait()
```

`stop_all()` is really just meant to wrap that into 1 call.

but, adding a thread into the shutdown mix would require a `wait_all()` func to ensure all cloudsyncs are shutdown.  it also adds complexity without a clear benefit.